engine/compiler/compiler.go

```diff
@@ -500,7 +500,9 @@ func (c *Compiler) Compile(ctx context.Context, args runtime.CompilerArgs) runti
 			// if the secret was already fetched and stored in the
 			// secret map it can be skipped.
 			if _, ok := spec.Secrets[s.Name]; ok {
-				step.SpecSecrets = append(step.SpecSecrets, spec.Secrets[s.Name])
+				if !isSecretPresent(step.SpecSecrets, spec.Secrets[s.Name]) {
```

Comment:
this is O(n^2). Although it would not make a difference but let's still try to keep methods as efficient as possible.
Create a map for step.SpecSecrets first and then check if it exists